articles/knsh14.re

```diff
@@ -0,0 +1,214 @@
+ウィンドウに画像や文字を出したり、キーの入力を受けたりする処理は別の@<code>{goroutine}で実行する必要があります。
+
+=== Hello world
+GopherConで例として利用されたHello worldのコード@<fn>{knsh14_gioui_sample_hello_world_link}を実行します。
+
+//list[knsh14_gioui_sample_hello_world][GioにおけるHello world][go]{
+package main
+
+import (
+  "gioui.org/ui"
+  "gioui.org/ui/app"
+  "gioui.org/ui/layout"
+  "gioui.org/ui/measure"
+  "gioui.org/ui/text"
+
+  "golang.org/x/image/font/gofont/goregular"
+  "golang.org/x/image/font/sfnt"
+)
+
+// START OMIT
+func main() {
+  go func() {
+    w := app.NewWindow()
+    regular, _ := sfnt.Parse(goregular.TTF)
+    var cfg ui.Config
+    var faces measure.Faces
+    ops := new(ui.Ops)
+    for e := range w.Events() {
+      if e, ok := e.(app.UpdateEvent); ok {
+        cfg = &e.Config
+        cs := layout.RigidConstraints(e.Size)
+        ops.Reset()
+        faces.Reset(cfg)
+
+        lbl := text.Label{Face: faces.For(regular, ui.Sp(72)), Text: "Hello, World!"} // HLdraw
+        lbl.Layout(ops, cs)                                                           // HLdraw
+
+        w.Update(ops)
+      }
+    } // HLeventloop
+  }()
+  app.Main()
+}
+
+// END OMIT
+//}
+
+このコードを実行すると@<img>{knsh14_gioui_hello_world}のような実行結果が得られます。
+//image[knsh14_gioui_hello_world][GioにおけるHello world]{
+//}
+
+ウィンドウを出す方法は真っ白なウィンドウを出す方法と変わりません。
+このサンプルでは文字を出す方法を新しく紹介します。
+23行目のfor文でウィンドウオブジェクトからイベントを取り出します。
+イベントには「画面を更新した」、「何らかの入力を受けた」などがあります。
+このサンプルでは画面を更新する際のイベント@<code>{gioui.org/ui/app.UpdateEvent}@<fn>{knsh14_gioui_app_updateevent_doc_link}の場合に文字を出す処理を行います。
+
+文字を出力するためには@<code>{gioui.org/ui/text.Label}を利用します。
+@<code>{Label}には2つのフィールドがあります。
+1つ目はもちろん出力するための@<code>{Text}フィールドです。
+2つ目は文字のフォント、大きさを決める@<code>{Face}フィールドです。
+これら2つを指定して文字を表示するための準備をします。
+
+文字を実際に出すためには@<code>{Layout}メソッドを呼び出す必要があります。
+このメソッドに渡すために、@<code>{gioui.org/ui.Ops}と@<code>{gioui.org/ui/layout.Constraints}を取得する必要があります。
+@<code>{gioui.org/ui.Ops}はループの外側で定義し、再利用することができます。
+@<code>{gioui.org/ui/layout.Constraints}はイベントを処理する毎に取得する必要があります。
+なぜなら画面サイズが変わったりした場合に再度計算する必要があるからです。
+
+//footnote[knsh14_gioui_sample_hello_world_link][https://github.com/eliasnaur/gophercon-2019-talk/blob/master/helloworld.go]
+//footnote[knsh14_gioui_app_updateevent_doc_link][https://godoc.org/gioui.org/ui/app#UpdateEvent]
+
+=== 画像を出してみる
+
+//list[knsh14_gioui_sample_display_image][Gioで画像を出す][go]
+package main
+
+// A simple Gio program. See https://gioui.org for more information.
+
+import (
+  "fmt"
+  "image"
+  _ "image/png"
+  "log"
+  "os"
+
+  "gioui.org/ui"
+  "gioui.org/ui/app"
+  "gioui.org/ui/layout"
+  "gioui.org/ui/measure"
+  "gioui.org/ui/widget"
+)
+
+func main() {
+  go func() {
+    w := app.NewWindow()
+    if err := loop(w); err != nil {
+      log.Fatal(err)
+    }
+  }()
+  app.Main()
+}
+
+func loop(w *app.Window) error {
+  var cfg app.Config
+  var faces measure.Faces
+  file, err := os.Open("gophercon.png")
+  defer file.Close()
+  if err != nil {
+    return err
+  }
+  img, _, err := image.Decode(file)
+  if err != nil {
```

Comment:
**[textlint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Disallow to use "!". (ja-technical-writing/no-exclamation-question-mark)